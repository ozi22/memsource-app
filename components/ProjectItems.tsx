import React, { FC } from 'react';
import { ScrollView, StyleSheet, TouchableOpacity, Text, View } from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import ProjectsDuck from '../redux/projects/projects.duck';
import UIDuck from '../redux/ui/ui.duck';
import ProjectsService from '../services/projects/projects.service';
import LoginDuck from '../redux/login/login.duck';
import { useNavigation } from '@react-navigation/native';
import Colors, { statuses } from '../constants/Colors';

const ProjectItems: FC = () => {
  const { content } = useSelector(ProjectsDuck.getProjects);
  const token: string = useSelector(LoginDuck.getUserToken);
  const dispatch = useDispatch();
  // TODO: add a type
  const navigation = useNavigation<any>();

  const getDetail = (id: string) => {
    dispatch(UIDuck.setLoading(1));
    dispatch(ProjectsDuck.clearDetail());
    ProjectsService.getDetail(token, id)
      .then(({ data }) => {
        dispatch(ProjectsDuck.setDetail(data));
      })
      .then(() => navigation.navigate('Detail'))
      .catch((e) => {
        // tslint:disable-next-line:no-console
        console.error(e);
      })
      .finally(() => dispatch(UIDuck.setLoading(-1)));
  };

  const renderItems = () => {
    return content?.map((item, index) => (
      <TouchableOpacity key={`${index}-project`} style={styles.itemContainer} onPress={() => getDetail(item.id)}>
        <View>
          <Text style={styles.itemName}>{item.name}</Text>
          <View style={{ flexDirection: 'row' }}>
            <Text style={styles.lang}>
              {item.sourceLang}
              {` > `}
              {item.targetLangs.map((item, index, { length }) => `${item}${index !== length - 1 ? ', ' : ''}`)}
            </Text>
          </View>
        </View>
        <Text style={[styles.status, { backgroundColor: statuses[item.status] }]}>{item.status}</Text>
      </TouchableOpacity>
    ));
  };
  return <ScrollView style={styles.container}>{renderItems()}</ScrollView>;
};

export default ProjectItems;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#f3f7f9',
    margin: 15
  },
  itemContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginBottom: 10,
    borderWidth: 1,
    borderStyle: 'solid',
    borderColor: '#dcdee2',
    backgroundColor: '#fff',
    padding: 10
  },
  itemName: {
    fontWeight: 'bold',
    color: Colors.light.text,
    fontSize: 16,
    paddingBottom: 3
  },
  lang: {
    display: 'flex',
    alignItems: 'center',
    color: Colors.light.tabIconDefault
  },
  status: {
    alignSelf: 'center',
    padding: 10
  }
});
