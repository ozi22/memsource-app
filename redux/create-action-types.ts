export const createActionTypes = (actionTypes: any = {}, duckName = 'DUCK'): any => {
  return Object.keys(actionTypes).reduce(
    (res, key) => ({
      ...res,
      [key]: `${duckName.toUpperCase()}_DUCK/${actionTypes[key].toUpperCase()}`
    }),
    {}
  );
};
