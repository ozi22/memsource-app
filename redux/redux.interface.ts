import { ThunkAction } from 'redux-thunk';

export interface IAction<Payload = any, Opt = any> {
  type: string;
  payload: Payload;
  opt?: Opt;
}

export type ThunkResult<R> = ThunkAction<R, any, undefined, any>;
