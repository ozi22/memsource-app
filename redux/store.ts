import { applyMiddleware, combineReducers, createStore } from 'redux';
import reduxThunk from 'redux-thunk';
import { composeWithDevTools } from 'redux-devtools-extension';
import UIDuck from './ui/ui.duck';
import LoginDuck from './login/login.duck';
import ProjectsDuck from './projects/projects.duck';

const combineDucks = (...ducks: any): any =>
  combineReducers(
    ducks.reduce((root: any, duck: any) => {
      return { ...root, [duck.name]: duck.reducer };
    }, {})
  );

export type AppState = ReturnType<typeof combineDucks>;

const rootReducer: any = combineDucks(UIDuck, LoginDuck, ProjectsDuck);

export const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(reduxThunk)));
