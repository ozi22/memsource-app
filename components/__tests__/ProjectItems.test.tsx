import React from 'react';
import renderer from 'react-test-renderer';
import ProjectItems from '../ProjectItems';
import { Provider } from 'react-redux';
import { store } from '../../redux/store';

jest.mock('@react-navigation/native');

describe('ProjectItems', () => {
  it('should render correctly', () => {
    const { toJSON } = renderer.create(
      <Provider store={store}>
        <ProjectItems />
      </Provider>
    );
    expect(toJSON()).toMatchSnapshot();
  });
});
