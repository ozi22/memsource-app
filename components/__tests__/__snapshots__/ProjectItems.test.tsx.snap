// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectItems should render correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#f3f7f9",
      "flex": 1,
      "margin": 15,
    }
  }
>
  <View />
</RCTScrollView>
`;
